src/core/shape_inference/include/gru_sequence_shape_inference.hpp

```diff
@@ -0,0 +1,178 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+#include <openvino/core/validation_util.hpp>
+#include <openvino/op/gru_sequence.hpp>
+
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+namespace rnn_seq {
+template <class OpType, class ShapeType>
+void validate_inputs_rank(const OpType* op,
+                          const std::vector<ShapeType>& input_shapes,
+                          const std::vector<Rank>& expected_ranks) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() >= expected_ranks.size(), "Can't validate inputs rank.");
+    for (auto i = 0; i < expected_ranks.size(); ++i) {
+        NODE_VALIDATION_CHECK(op,
+                              input_shapes[i].rank().compatible(expected_ranks[i]),
+                              "Shape rank of input at ",
+                              i,
+                              " is incompatible. Expected rank: ",
+                              expected_ranks[i],
+                              ", actual shape: ",
+                              input_shapes[i],
+                              ".");
+    }
+}
+
+// Output shapes layout:
+// output_shapes[0]: [batch_size, num_directions, seq_length, hidden_size] // Rank always 4
+// output_shapes[1]: [batch_size, num_directions, hidden_size] // Rank always 3
+template <class OpType, class ShapeType>
+void gru_shape_infer(const OpType* op,
+                     const std::vector<ShapeType>& input_shapes,
+                     std::vector<ShapeType>& output_shapes) {
+    NODE_VALIDATION_CHECK(op,
+                          input_shapes.size() >= 6 && output_shapes.size() == 2,
+                          "Incorrect number of shapes has been provided.");
+
+    auto& y_out_shape = output_shapes[0];
+    auto& ho_out_shape = output_shapes[1];
+    y_out_shape.resize(4);   // Rank always 4
+    ho_out_shape.resize(3);  // Rank always 3
+
+    rnn_seq::validate_inputs_rank(op, input_shapes, {3, 3, 1, 3, 3, 2});
+
+    auto x_pshape = input_shapes[0];
+    auto ht_pshape = input_shapes[1];
+    auto sl_pshape = input_shapes[2];
+    auto w_pshape = input_shapes[3];
+    auto r_pshape = input_shapes[4];
+    auto b_pshape = input_shapes[5];
+
+    using DimType = typename std::iterator_traits<typename ShapeType::iterator>::value_type;
+
+    // Merge batch_size dimension across all inputs to evaluate output[0] dimension
+    DimType merged_batch_size = x_pshape.rank().is_static() ? x_pshape[0] : DimType();
```

Comment:
echo: better return immediately instead of use "DimType()" to propagate.

Reply:
If one input has dynamic rank, it doesn't mean we can't propagate some information about dimensions from other inputs.
Also dynamic rank case shouldn't be possible when `shape_infer` is called by plugins with `StaticShape` as template type.